### PR TITLE
Issue #3467: Replaced .coordinates() with .positions

### LINKED ISSUE #3467: replace atoms.coordinates() with atoms.positions
Since 2.0, atoms does not have coordinates() anymore and only positions is supported.

- [ ] Needs to be fixed:
https://github.com/MDAnalysis/mdanalysis/blob/46c8badd8b8967dc25f8125ef87582490ab497e3/package/MDAnalysis/analysis/encore/covariance.py#L218
- [ ] Also, fix benchmarks that still use coordinates() (but this could be made another PR)

### PR DESCRIPTION
**Pull request by pranjii**
Fixes MDAnalysis/mdanalysis#3467 

Changes made in this Pull Request:
 -Replaced .coordinates() with .positions

PR Checklist
------------
 - [ ] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

**Comment by pranjii** [comment]
@orbeckst Hey one of my tests seem to be failing could you help me out fixing it?

File: package/MDAnalysis/analysis/encore/covariance.py [hunk 1]
```diff
@@ -215,7 +215,7 @@ def covariance_matrix(ensemble,
         # Select the same atoms in reference structure
         reference_atom_selection = reference.select_atoms(
             ensemble.get_atom_selection_string())
-        reference_coordinates = reference_atom_selection.atoms.coordinates()
+        reference_coordinates = reference_atom_selection.atoms.positions
```
Comment (IAlibay) [inline package/MDAnalysis/analysis/encore/covariance.py:218]: The failed codecov check is indicating that the this line isn't covered by a test. This explains why tests didn't identify that `atoms.coordinates()` no longer existed.

If possible could you add a test that checks this code path?

**Review by IAlibay — CHANGES_REQUESTED** [review]
Thanks for opening this PR @pranjii 

The failing test is essentially an indication that there is a section of the code not covered by tests. Ideally we would want to add a test here to make sure that the code is working as expected.

Please also do add yourself to AUTHORS and update the CHANGELOG accordingly.

**Comment by pranjii** [comment]
> Thanks for opening this PR @pranjii
> 
> The failing test is essentially an indication that there is a section of the code not covered by tests. Ideally we would want to add a test here to make sure that the code is working as expected.
> 
> Please also do add yourself to AUTHORS and update the CHANGELOG accordingly.

working on it

**Comment by pranjii** [comment]
@IAlibay  Seems like I am not the guy who can do the tests. Should I revert my commits or will the PR be closed?

**Comment by IAlibay** [comment]
> @IAlibay Seems like I am not the guy who can do the tests. Should I revert my commits or will the PR be closed?

Can you go into more details? If you want to walk us through what you think is going wrong I'm sure we can do our best to guide you towards tests that pass.

So this is your current error `E       AttributeError: module 'MDAnalysis.analysis.encore' has no attribute 'covariance_matrix'`. Can you try to write a comment here to explain what you're trying to do and then we can start from there?

**Comment by orbeckst** [comment]
>  Seems like I am not the guy who can do the tests. 

I approved running the tests — if this is what you mean?

We would want you to write actual tests for MDAnalysis that are run to check your fix. Have you written tests with pytests before? See https://userguide.mdanalysis.org/stable/testing.html to learn more about tests and to get started.

**Comment by pranjii** [comment]
@IAlibay Thank u. I will probably need some time to figure it all out and thank u for offering ur help. Right now I have a question:

I didn't see any of the functions in `covariance.py` tested. Is it supposed to be so? If yes is the location right now ok or should is there some other location you would prefer? 

Right now the problem is that I was unaware of how advanced was the mathematical concept I was dealing with, the  `E AttributeError: module 'MDAnalysis.analysis.encore' has no attribute 'covariance_matrix'. ` was a totally silly error that I made during the pep8 changes and due to my lack of experience in testing functions but I will work on each of the issue and will deliver the final product.

Thank u for the support.

**Comment by orbeckst** [comment]
Hello @pranjii , are you still working on this PR?

**Comment by IAlibay** [comment]
> I didn't see any of the functions in covariance.py tested. Is it supposed to be so? If yes is the location right now ok or should is there some other location you would prefer?

Sorry I failed to respond to this. The tests should be under `testsuite/MDAnalysisTests/analysis/test_encore.py`.

**Comment by ojeda-e** [comment]
@orbeckst @IAlibay I noticed that lines [216-217](https://github.com/MDAnalysis/mdanalysis/blame/3e249fe7173e68bb61d4ad7a2dfd316014760410/package/MDAnalysis/analysis/encore/covariance.py#L217) in `covariance.py` have no previous coverage, then a brand new unit test should be added. However, these lines are a bit problematic because of two main reasons:
1. `get_atom_selection_string()` is not an attribute of Universe.
2. It seems like the if statement doesn't do what the docstrings indicate. With no reference, it should run a covariance calculation using distance to the mean. I may be wrong, but shouldn't it be something like this?
```python3
    if reference is not None:
        reference_coordinates = reference.atoms.positions.flatten()
    else:
        # Covariance calculation using distance to the mean
        reference_coordinates = coordinates - np.mean(coordinates, axis=0)
```
Additionally, tests for similarity uses covariance in `test_encore.py`, and if I am correct about item (2), fixing it would mean that additional unit tests should be modified because `covariance_matrix`, `ml_covariance_estimator`,  and `shrinkage_covariance_estimator` are used to calculate harmonic ensemble similarity, hence there is a possibility that HES tests will fail. 
If I am correct I could help with this issue. Otherwise, I'd appreciate if you let me know what I am missing here, thanks.

**Comment by orbeckst** [comment]
Is ensemble really a Universe or something else (even though the docs say Universe object)?

> Am 1/13/22 um 01:10 schrieb Estefania Barreto-Ojeda ***@***.***>:
> 
> get_atom_selection_string() is not an attribute of Universe.

**Comment by ojeda-e** [comment]
> Is ensemble really a Universe or something else (even though the docs say Universe object)?

Yes, an ensemble is a Universe defined with `timeseries` as a bound method. For example:

```Python3
from MDAnalysisTests.datafiles import DCD, PSF
import MDAnalysis as mda

traj = mda.Universe(PSF, DCD)
ens1 = mda.Universe(PSF, traj.trajectory.timeseries())
```

I don't know if the `get_atom_selection_string()` was an attribute of Universe in previous versions of MDAnalysis, but I can't find it anywhere else in the library. 
Since an ensemble is a Universe, I guess lines [216-217 ](https://github.com/MDAnalysis/mdanalysis/blame/3e249fe7173e68bb61d4ad7a2dfd316014760410/package/MDAnalysis/analysis/encore/covariance.py#L217)are meant to be:
```Python3
reference_atom_selection = ensemble.select_atoms(reference)
```
instead of
```Python3
reference_atom_selection = reference.select_atoms(
            ensemble.get_atom_selection_string())
```
If I am understanding correctly. (Spoiler alert: I may not)

**Comment by jbarnoud** [comment]
From what I can tell, the call to `get_atom_selection_string` was introduced in 5d914cc0b6186c6ab550a6690ee56438983b72f1 and at no point was the method defined. It looks to me that this part of the code never worked and is probably a leftover from when ENCORE was developed outside of MDAnalysis. Maybe @mtiberti has an idea about it?

**Comment by jbarnoud** [comment]
Fixing  ENCORE is out of scope for this PR. I opened MDAnalysis/mdaencore#29 and MDAnalysis/mdanalysis#3539 to track that issue.

Regarding the original issue, the code about rms is preceded by a comment that says:

```
# ag.positions is the new syntax
# but older commit hashes will need to use
# ag.coordinates()
```

We run daily benchmarks, the result of which are visible here: <https://www.mdanalysis.org/benchmarks/>. The benchmarks need to be ale to run on old commits as well, some of which being older than the `positions` attribute. Hence the try/except. We recently changed the minimum commit we benchmark to a version that do have the attribute (<https://github.com/MDAnalysis/benchmarks/pull/12#issuecomment-1031629405>) so the try/except is not useful anymore and became dead code. In this instance, replacing `.coordinates()` by `.positions` means repeating the same code twice. Instead, we should remove the dead code.

The next step would be to:
* [ ] remove the dead code in the rms benchmark
* [ ] remove the broken test, it will be added back when we know how to fix the broken code
* [ ] Update the CHANGELOG
* [ ] Add yourself to the AUTHOR file

**Comment by jbarnoud** [comment]
By the way, MDAnalysis is currently in the process of evaluating a potential change in its license. As part of this we are asking that all new-ish contributors introduce themselves (with an email address that they regularly monitor) to the [MDAnalysis developer list](https://groups.google.com/group/mdnalysis-devel) before we accept their contributions. Please include your github handle and a link to this pull request. The aim is that we then have a means of contacting the developers should we decide to change our license in the near future.

Thank you.

**Comment by orbeckst** [comment]
Hello @pranjii are you still working on this PR?

There hasn't been any activity (including questions) for over a month so unless we hear anything over the next 4 days we will [close this PR as stale](https://github.com/MDAnalysis/mdanalysis/wiki/GSoC-FAQ#closing-stale-prs).

**Comment by orbeckst** [comment]
I am closing this PR due to inactivity.

PR MDAnalysis/mdanalysis#3576 is now considered primary PR for issue MDAnalysis/mdanalysis#3467.

**Pull request closed without merging**